### PR TITLE
Promtail: Fix deadlock on tailer shutdown.

### PR DESCRIPTION
**Pull request by slim-bean**
to avoid a deadlock on shutdown, leave the goroutine running which reads the Lines channel from the tailer and it will exit when the tailer closes the channel.

Also updated the underlying hp tail fork we have to return an os.ErrNotExist when you call the Size or Tell methods and the tailer is not currently tailing a file (happens when a file is deleted before we have a chance to remove the tailer or on file rotation)

### CODE REVIEW

File: pkg/promtail/targets/file/tailer.go [hunk 1]
```diff
@@ -136,31 +136,38 @@ func (t *tailer) markPositionAndSize() error {
...
 	defer t.posAndSizeMtx.Unlock()
 
-	pos, err := t.tail.Tell()
+	size, err := t.tail.Size()
 	if err != nil {
+		// If the file no longer exists, no need to save position information
+		if err == os.ErrNotExist {
+			level.Info(t.logger).Log("msg", "skipping update of position for a file which does not currently exist")
+			return nil
+		}
```
Comment (slim-bean) [inline pkg/promtail/targets/file/tailer.go:145]: This is really the only important change in this function, the diff is screwy because I swapped the order of Tell() and Size() calls.
This is a more graceful way to handle ignoring the size/position for a file which doesn't exist, it leverages the change in the HP tail lib which returns the os.ErrNotExist error

File: pkg/promtail/targets/file/tailer.go [hunk 2]
```diff
@@ -136,31 +136,38 @@ func (t *tailer) markPositionAndSize() error {
...
 		return err
 	}
-	totalBytes.WithLabelValues(t.path).Set(float64(size))
+	readBytes.WithLabelValues(t.path).Set(float64(pos))
+	t.positions.Put(t.path, pos)
 
 	return nil
 }
 
-func (t *tailer) stop(removed bool) {
```
Comment (slim-bean) [inline pkg/promtail/targets/file/tailer.go:155]: In a change I made recently, I added this bool to do the conditional position update below but that isn't necessary (and was also not implemented correctly), it can be removed, calling the markPositionAndSize is basically a NOOP now if the file does not exist

File: pkg/promtail/targets/file/tailer.go [hunk 3]
```diff
@@ -136,31 +136,38 @@ func (t *tailer) markPositionAndSize() error {
...
-		err := t.markPositionAndSize()
-		if err != nil {
-			level.Error(t.logger).Log("msg", "error marking file position when stopping tailer", "path", t.path, "error", err)
-		}
+	err := t.markPositionAndSize()
+	if err != nil {
+		level.Error(t.logger).Log("msg", "error marking file position when stopping tailer", "path", t.path, "error", err)
+	}
+
+	err = t.tail.Stop()
```
Comment (slim-bean) [inline pkg/promtail/targets/file/tailer.go:167]: Stop the tail here, these leaves the main goroutine running still which will consume the tail.Lines channel until it's closed

File: pkg/promtail/targets/file/tailer.go [hunk 4]
```diff
@@ -105,12 +98,21 @@ func (t *tailer) run() {
...
-				level.Error(t.logger).Log("msg", "error getting tail position and/or size, stopping tailer", "path", t.path, "error", err)
+				level.Error(t.logger).Log("msg", "position timer: error getting tail position and/or size, stopping tailer", "path", t.path, "error", err)
+				// To prevent a deadlock on stopping the tailer we need to launch a thread to consume any unread lines
+				go func() {
+					for range t.tail.Lines {}
+				}()
+				t.tail.Stop()
+				if err != nil {
+					level.Error(t.logger).Log("msg", "position timer: error stopping tailer", "path", t.path, "error", err)
+				}
```
Comment (slim-bean) [inline pkg/promtail/targets/file/tailer.go:109]: This is to help recover from weird errors we have seen in some environments, if we fail to update the position information on the timer we want to close the tailer so it can be re-opened by the upper level sync function.

Adding the goroutine here to consume t.tail.Lines makes sure we dont' deadlock on lines being in the channel when we tel the tail to stop

File: pkg/promtail/targets/file/tailer.go [hunk 5]
```diff
@@ -125,8 +127,6 @@ func (t *tailer) run() {
 			if err := t.handler.Handle(model.LabelSet{}, line.Time, line.Text); err != nil {
 				level.Error(t.logger).Log("msg", "error handling line", "path", t.path, "error", err)
 			}
-		case <-t.quit:
-			return
```
Comment (slim-bean) [inline pkg/promtail/targets/file/tailer.go:129]: Removing the quit channel and instead we will quit when the underlying tailer closes the t.tail.Lines channel above.

File: pkg/promtail/targets/file/tailer.go [hunk 6]
```diff
@@ -105,12 +98,21 @@ func (t *tailer) run() {
 		case <-positionWait.C:
 			err := t.markPositionAndSize()
 			if err != nil {
-				level.Error(t.logger).Log("msg", "error getting tail position and/or size, stopping tailer", "path", t.path, "error", err)
+				level.Error(t.logger).Log("msg", "position timer: error getting tail position and/or size, stopping tailer", "path", t.path, "error", err)
+				// To prevent a deadlock on stopping the tailer we need to launch a thread to consume any unread lines
+				go func() {
+					for range t.tail.Lines {}
+				}()
```
Comment (rfratto) [inline pkg/promtail/targets/file/tailer.go:105]: I'm finding it a little hard to follow the logic here, unfortunately. I think this could benefit from some comments around the lifecycle of this loop. I guess the way it works now is that one of the two select cases will eventually fail/be closed? 

Also, even though you only need the goroutine to drain the lines channel in this specific branch, IMO it'd be more resilient to future changes if it was drained in a defer to the whole run loop.

File: pkg/promtail/targets/file/tailer.go (hunk 1 above)
Comment (rfratto) [inline pkg/promtail/targets/file/tailer.go:145]: Small nit: can you make this an if/else if?

```
size, err := t.tail.Size()
if err == os.ErrNotExist {
  // ...
  return nil
} else if err != nil {
  return err
} 
```

**Review by rfratto — APPROVED** [review]
LGTM! I just have small nits, but I'm concerned about the vendor update. There's a few more cases where `tail.file` is being modified or referenced without holding the lock on the new mutex. I think we should play it safe and protect all accesses to it.

File: pkg/promtail/targets/file/tailer.go [hunk 7]
```diff
@@ -136,33 +162,66 @@ func (t *tailer) markPositionAndSize() error {
...
+		// yielding the element type's zero value after any previously sent values have been received.
+		return
+	default:
+		// Also from the spec: If one or more of the communications can proceed,
+		// a single one that can proceed is chosen via a uniform pseudo-random selection.
+		// Otherwise, if there is a default case, that case is chosen.
+		// We have no sender on t.stopped so that case can only proceed when the channel is closed.
+		// If we hit the default case it's because nobody has closed the channel and therefore nobody has called stop.
+		break
+	}
```
Comment (rfratto) [inline pkg/promtail/targets/file/tailer.go:204]: Another way of doing this would be to wrap a call to `stop` with a `sync.Once`, which would also let you get rid of `stopMtx`. It's not clear to me if either is "better" but using the `sync.Once` would probably need less explanation in the code.

File: vendor/github.com/hpcloud/tail/tail.go [hunk 8]
```diff
@@ -191,7 +201,9 @@ func (tail *Tail) close() {
 func (tail *Tail) closeFile() {
 	if tail.file != nil {
 		tail.file.Close()
+		tail.fileMtx.Lock()
 		tail.file = nil
+		tail.fileMtx.Unlock()
```
Comment (rfratto) [inline vendor/github.com/hpcloud/tail/tail.go:206]: Unfortunately I think this whole function needs to hold the lock, not just when you set `tail.file` to nil

**Review by rfratto — APPROVED** [review]
LGTMBWN! (Looks Good To Me But With Nits)

Thanks for making the changes, I think this is a lot easier to follow. Now that I think about it, doing it this way also guarantees we'll finish writing all lines to Loki, which I'm not sure would've happened with your first attempt at this.

File: vendor/github.com/hpcloud/tail/tail.go (hunk 8 above)
Comment (slim-bean) [reply vendor/github.com/hpcloud/tail/tail.go]: good catch, fixed

File: pkg/promtail/targets/file/tailer.go (hunk 7 above)
Comment (slim-bean) [reply pkg/promtail/targets/file/tailer.go]: good feedback, done.
